aiosip/peers.py

```diff
@@ -85,35 +86,39 @@ def _create_dialog(self, method, from_details, to_details, contact_details=None,
 
     async def request(self, method, from_details, to_details, contact_details=None, password=None, call_id=None,
                       headers=None, cseq=0, payload=None, dialog_factory=Dialog, timeout=None, **kwargs):
+        from .message import Request
+        from .transaction import start_client_transaction
 
-        dialog = self._create_dialog(method=method,
-                                     from_details=from_details,
-                                     to_details=to_details,
-                                     contact_details=contact_details,
-                                     headers=headers,
-                                     payload=payload,
-                                     password=password,
-                                     call_id=call_id,
-                                     cseq=cseq,
-                                     dialog_factory=dialog_factory,
-                                     **kwargs)
-        try:
-            response = await dialog.start(timeout=timeout)
-            dialog.status_code = response.status_code
-            dialog.status_message = response.status_message
-            return dialog
-        except asyncio.CancelledError:
-            dialog.cancel()
-            raise
+        if not contact_details:
+            host, port = self.local_addr
 
-    async def subscribe(self, expires=3600, **kwargs):
+            if self._app.defaults['override_contact_host']:
+                host = self._app.defaults['override_contact_host']
+            elif host == '0.0.0.0' or host.startswith('127.'):
+                host = from_details['uri']['host']
 
-        if expires:
-            headers = kwargs.get('headers', CIMultiDict())
-            headers['Expires'] = expires
-            kwargs['headers'] = headers
+            contact_details = Contact(
+                {
+                    'uri': 'sip:{username}@{host_and_port};transport={protocol}'.format(
+                        username=from_details['uri']['user'],
+                        host_and_port=utils.format_host_and_port(host, port),
+                        protocol=type(self._protocol).__name__.lower()
+                    )
+                })
 
-        return await self.request('SUBSCRIBE', **kwargs)
+        message = Request(method, 20, from_details, to_details, contact_details, headers, payload)
+        # TODO: this is weird
+        return await start_client_transaction(self._app, message, self)
+
+    async def subscribe(self, expires=3600, **kwargs):
```

Comment:
This seems to aim to be a high level subscribe method. If so, I'd expect it to return a 'subscription' object or a 'dialog' if you want to go a bit lower level, but not a transaction.

Reply:
I've been hunting for python-compatible SIP APIs and just today I stumbled upon this:

http://sipsimpleclient.org/projects/sipsimpleclient/wiki/SipCoreApiDocumentation#Subscription

That API is exactly what I had in mind with my previous comment. You may take some inspiration from that API as well.